### PR TITLE
NEW: Add Riemmanian minimum distance to mean (MDM) classification for SPD matrices

### PR DESCRIPTION
**Pull request by ninamiolane**
This PR follows the stalled PR #623

### CODE REVIEW

File: geomstats/geometry/special_orthogonal.py [hunk 1]
```diff
@@ -112,6 +120,101 @@ def random_uniform(self, n_samples=1, tol=1e-6):
         rotation_mat, _ = gs.linalg.qr(random_mat)
         return rotation_mat
 
+    def rotation_from_angle(self, theta):
```
Comment (nguigs) [inline geomstats/geometry/special_orthogonal.py:123]: Isn't there a dedicated method `rotation_matrix_from_vector`?

File: geomstats/geometry/special_orthogonal.py [hunk 2]
```diff
@@ -112,6 +120,101 @@ def random_uniform(self, n_samples=1, tol=1e-6):
...
+
+        Returns
+        -------
+        array-like, shape=[2, 2]
+            2D rotation with divided angle.
+        """
+        if self.n != 2:
+            raise NotImplementedError
+        angle = gs.einsum(
+            '...,...->...', self.angle_of_rot2(point), alpha)
```
Comment (nguigs) [inline geomstats/geometry/special_orthogonal.py:189]: idem

File: geomstats/geometry/special_orthogonal.py [hunk 3]
```diff
@@ -680,6 +782,127 @@ def regularize_tangent_vec(
...
+
+        aux_mat = gs.einsum(
+            'nij,njk->nik', mat_unitary_u, new_mat_diag_s)
+        rot_mat += gs.einsum(
+            'n,njk->njk', mask_float,
+            gs.einsum('nij,njk->nik', aux_mat, mat_unitary_v))
+        return rot_mat
+
+    @geomstats.vectorization.decorator(['else', 'vector'])
+    def skew_matrix_from_vector(self, vec):
```
Comment (nguigs) [inline geomstats/geometry/special_orthogonal.py:819]: These are not necessary anymore

File: geomstats/geometry/special_orthogonal.py [hunk 4]
```diff
@@ -680,6 +782,127 @@ def regularize_tangent_vec(
...
+        cross_prod_1 = gs.to_ndarray(cross_prod_1, to_ndim=3, axis=1)
+        cross_prod_2 = gs.to_ndarray(cross_prod_2, to_ndim=3, axis=1)
+        cross_prod_3 = gs.to_ndarray(cross_prod_3, to_ndim=3, axis=1)
+        skew_mat = gs.concatenate(
+            [cross_prod_1, cross_prod_2, cross_prod_3], axis=1)
+        return skew_mat
+
+    @staticmethod
+    @geomstats.vectorization.decorator(['matrix', 'output_point'])
+    def vector_from_skew_matrix(skew_mat):
```
Comment (nguigs) [inline geomstats/geometry/special_orthogonal.py:885]: same

File: geomstats/visualization.py [hunk 5]
```diff
@@ -647,3 +663,86 @@ def plot(points, ax=None, space=None,
...
+        self.fig = fig
+
+    def draw(self, x_from, y_from, x_to, y_to):
+        """
+        Draw the arrow in data coordinates.
+
+        :param x_from: x coordinate at arrow base
+        :param y_from: y coordinate at arrow base
+        :param x_to: x coordinate at arrow head
+        :param y_to: y coordinate at arrow head
```
Comment (nguigs) [inline geomstats/visualization.py:681]: Not our usual docstring syntax

File: tests/test_spd_matrices.py [hunk 6]
```diff
@@ -75,6 +88,40 @@ def test_random_point_and_belongs_vectorization(self):
         expected = gs.array([True] * 4)
         self.assertAllClose(result, expected)
 
+    @geomstats.tests.np_only
```
Comment (nguigs) [inline tests/test_spd_matrices.py:91]: Could we check other backends?

File: geomstats/geometry/special_orthogonal.py [hunk 7]
```diff
@@ -112,6 +120,101 @@ def random_uniform(self, n_samples=1, tol=1e-6):
...
+        mat_first_line = gs.transpose(gs.array([c, -s]))
+        mat_second_line = gs.transpose(gs.array([s, c]))
+        mat_first_line = gs.expand_dims(
+            mat_first_line, -2)
+        mat_second_line = gs.expand_dims(mat_second_line, -2)
+        mat = gs.concatenate(
+            [mat_first_line, mat_second_line], axis=-2)
+        return mat
+
+    def angle_of_rot2(self, point):
```
Comment (nguigs) [inline geomstats/geometry/special_orthogonal.py:149]: likewise, this should be in `vector_from_rotation_matrix` for consistency

File: geomstats/geometry/special_orthogonal.py [hunk 8]
```diff
@@ -112,6 +120,101 @@ def random_uniform(self, n_samples=1, tol=1e-6):
...
+            Angle of 2D rotation.
+        """
+        if self.n != 2:
+            raise NotImplementedError
+        sinus = point[..., 1, 0]
+        cosinus = point[..., 0, 0]
+        return gs.einsum(
+            '...,...->...', gs.sign(sinus), gs.arccos(cosinus))
+
+    def multiply_angle_of_rot2(self, point, alpha):
```
Comment (nguigs) [inline geomstats/geometry/special_orthogonal.py:169]: we can generalise to dim n using exp/log

**Review by nguigs — COMMENTED** [review]
There are a few changes from the past that are not implemented here, and I some method can be generalized to arbitrary dim for consistency

File: geomstats/visualization.py [hunk 9]
```diff
@@ -26,6 +27,21 @@
                'H2_poincare_disk', 'H2_poincare_half_plane', 'H2_klein_disk',
                'poincare_polydisk']
 
+EPS = 1e-8
+
+
+def nice_matplotlib():
```
Comment (nguigs) [inline geomstats/visualization.py:33]: Nice is subjective, maybe use another name. Is this method necessary for the 2d ellipses only?

**Comment by nguigs** [comment]
Also there is a file result.png that should be removed

File: geomstats/geometry/special_orthogonal.py (hunk 7 above)
Comment (nguigs) [reply geomstats/geometry/special_orthogonal.py:149]: What do you think about having `n=2` specfic methods in `SO(n)`? Is there an alternative?

**Comment by ninamiolane** [comment]
I will extract the visualization Ellipsis2D from this in a new PR.

**Pull request closed without merging**